### PR TITLE
fix: ensure lxd vms with multiple nics are supported

### LINKED ISSUE #20440: Can not use several spaces when using LXD as substrate for Juju
### Description

Juju is failing to spawn a VM belonging to 2 spaces.

The following error is then shown in juju status' output and in the logs:
"ERROR juju.worker.provisioner cannot start instance for machine "0": Failed start validation for device "eth0": Instance DNS name "juju-8554ac-0" conflict between "eth0" and "eth1" because both are connected to same network" 

You can take  a look at the attached files for a sample 
Image

### Juju version

3.6.10

### Cloud

LXD

### Expected behaviour

Juju should succeed in deploying the LXD-backed VM and in correctly configuring it.

### Reproduce / Test

1. sudo lxc launch ubuntu:24.04 pg16 --vm 	-c limits.cpu=4 -c limits.memory=16GiB -d root,size=32GiB
(Note , I also tried to run the same steps inside a baremetal machine to avoid nested virtualization but still got same errors)
2. sudo lxc exec pg16 bash
3. su -l ubuntu
4. sudo apt -y update && sudo apt -y upgrade
5. export JUJU_VERSION=3.6/stable # I also tried with edge (version = 3.6.10-610241d, revision =  32234)
6. sudo snap install juju --channel=${JUJU_VERSION}
7. mkdir -p /home/ubuntu/.local/share
8. export LXD_VERSION=5.21/stable # I also tried with edge
9. sudo snap install lxd --channel=${LXD_VERSION}
10. sudo lxd init --auto
11. sudo lxc network set lxdbr0 ipv6.address none
12. sudo lxc network create lxdbr1 ipv4.address=192.168.100.1/24 ipv6.address=none
13. sudo lxc network create lxdbr2  ipv4.address=192.168.200.1/24  ipv6.address=none
14. juju bootstrap lxd test --bootstrap-constraints="virt-type=virtual-machine mem=4G"
15. juju add-model site1
16. juju subnets
17. juju spaces
18. juju add-space peers
19. juju move-to-space peers 192.168.100.0/24
20. juju add-space clients
21. juju move-to-space clients 192.168.200.0/24
22. export PG_VERSION=16/stable
23. export UBUNTU_BASE=24.04
24. juju deploy postgresql db1	--channel ${PG_VERSION} \
				--config profile=production \
--constraints "spaces=clients,peers virt-type=virtual-machine mem=4G " \
--bind="database-peers=peers database=clients" \
--base ubuntu@${UBUNTU_BASE}

### Notes & References

Please note that I tried to modify the bridges to disable dns by setting:
sudo lxc network set <bridge> dns.mode none
Yet, Juju was indefinitely waiting for the VM to get ready (despite being shown as running by sudo lxc list)

Please also note that Juju behaves correctly when using LXD as a container provider.
Please also note that I tried with both stable and edge versions of both LXD and Juju without any noticeable change

### PR DESCRIPTION
**Pull request by wallyworld**
The netplan config for lxd vms was missing directives to match on hw address. This is needed because the nic naming 
scheme on vms is enpXsp0 not ethX.

There's a fair bit of code duplication between the lxd provider and lxd container packages. The fix uses the same implementation as for the containers. Ideally we'd look to refactor the code, but not for this PR.

Also, we don't want more than one dhcp configured nic to get a default route. If this happens, it can prevent network traffic egress which is fixed by manually deleting the offending route entries. So add the extra netplan override to prevent it. Ideally there'd be a way to model what nic gets egress and/or configure the routes for each nic but the netplan we write for lxd is quite rigid. ie we just write a really simple netplan with `dhcp4: true` and don't provide a way to model/configure other aspects that I can see (apart from the matching algorithm).

**Edit:** after doing this PR, I saw this issue
https://github.com/juju/juju/issues/20540
which looks like the same issue I ran into, ie multiple default routes preventing egress
From the bug:
```
$ ip route show default
default via 192.168.200.1 dev eth2 proto dhcp src 192.168.200.133 metric 100 
default via 192.168.100.1 dev eth3 proto dhcp src 192.168.100.110 metric 100 
default via 10.175.141.1 dev eth0 proto dhcp src 10.175.141.144 metric 100 
default via 10.175.141.1 dev eth1 proto dhcp src 10.175.141.73 metric 100 
```
In the absence of a modelled approach, I just chose to allow a default route for the first dhcp enabled nic only, which I think is eth0.

## QA steps

Create 2 additional lxd networks besides the lxdbr0 one.
```
lxc network create lxdbr1 ipv4.address=10.68.48.1/24 ipv6.address=none
lxc network create lxdbr2 ipv4.address=10.68.49.1/24 ipv6.address=none
```

Add one of the above to the default profile to create conditions for a corner case.
```
devices:
  eth0:
    nictype: bridged
    parent: lxdbr0
    type: nic
  eth1:
    nictype: bridged
    parent: lxdbr1
    type: nic
```

Test with lxd container and vm. For the vm test, do the same steps but with additional constraints `virt-type=virtual-machine mem=4G`

```
juju bootstrap lxd test
juju switch controller
juju add-space clients 10.68.48.0/24
juju add-space peers 10.68.49.0/24
juju deploy postgresql --channel 16/stable --constraints "spaces=clients,peers " --bind="database-peers=peers database=clients" 
```
Check postgres comes up and is active.
Check netplan
eg container case
```
cat /etc/netplan/99-juju.yaml 
network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
    eth1:
      dhcp4: true
      dhcp4-overrides:
        use-routes: false
    eth2:
      dhcp4: true
      dhcp4-overrides:
        use-routes: false
```
eg vm case
```
cat /etc/netplan/99-juju.yaml 
network:
  version: 2
  ethernets:
    eth0:
      match:
        macaddress: 00:16:3e:e5:a3:e9
      dhcp4: true
    eth1:
      match:
        macaddress: 00:16:3e:5a:29:c7
      dhcp4: true
      dhcp4-overrides:
        use-routes: false
    eth2:
      match:
        macaddress: 00:16:3e:5c:08:b0
      dhcp4: true
      dhcp4-overrides:
        use-routes: false
```

## Links

**Issue:** Fixes #20440.

**Jira card:** [JUJU-8421](https://warthogs.atlassian.net/browse/JUJU-8421)

[JUJU-8421]: https://warthogs.atlassian.net/browse/JUJU-8421?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: container/lxd/network.go [hunk 1]
```diff
@@ -271,6 +271,19 @@ func isValidNetworkType(net *api.Network) bool {
 	return net.Type == netTypeBridge || net.Type == nicTypeMACVLAN
 }
 
+// NetworkName returns the network name from the
+// map of device attributes.
+func NetworkName(info map[string]string) string {
+	// Versions of the LXD profile prior to 3.22 have the network name as
+	// "parent" under NIC entries in the "devices" list.
+	// Later versions have it under "network".
```
Comment (SimonRichardson) [inline container/lxd/network.go:279]: TIL'd.

**Review by SimonRichardson — APPROVED** [review]
This feels like a valid approach for now.

**Review by manadart — COMMENTED** [review]
We're trying to solve https://github.com/juju/juju/issues/20540 in a drive-by with a fix for https://github.com/juju/juju/issues/20440, but we're a miss on the latter.

Let's stick with a single bug the patch.

Your reproduction scenario does not address falsifiability. By using `--bind="database-peers=peers database=clients"`, you are still effectively binding _all the other endpoints_ to the default, which remains `alpha`, so ending up with 3 NICs is correct.

Take another look at my reproduction of the issue. You don't need any profile changes, and you don't need VMs. As long as you have multiple LXD-managed bridges, you can reproduce by carving out some spaces and deploying with a bind stanza that should yield 2 NICs, but gets 3.

This behaviour is not addressed in the patch. 

- Add one extra bridge.
- Bootstrap, and create 2 spaces, one with the subnet from the default bridge, another with the new one.
- Deploy Postgres with the first space as the _default_ binding, and the second for one of the endpoints.
- We should get 2 NICs, one for each bound space.

**Comment by wallyworld** [comment]
> We're trying to solve #20540 in a drive-by with a fix for #20440, but we're a miss on the latter.
> 
> Let's stick with a single bug the patch.
> 
> Your reproduction scenario does not address falsifiability. By using `--bind="database-peers=peers database=clients"`, you are still effectively binding _all the other endpoints_ to the default, which remains `alpha`, so ending up with 3 NICs is correct.
> 
> Take another look at my reproduction of the issue. You don't need any profile changes, and you don't need VMs. As long as you have multiple LXD-managed bridges, you can reproduce by carving out some spaces and deploying with a bind stanza that should yield 2 NICs, but gets 3.
> 
> This behaviour is not addressed in the patch.
> 
>     * Add one extra bridge.
> 
>     * Bootstrap, and create 2 spaces, one with the subnet from the default bridge, another with the new one.
> 
>     * Deploy Postgres with the first space as the _default_ binding, and the second for one of the endpoints.
> 
>     * We should get 2 NICs, one for each bound space.

In my testing I suspect I came across a different issue that was triggered by testing with vm as per the bug repro steps. The root cause was the missing mac address match stanza in netplan for the lxd provider (we were only using that for lxd vms deployed to machine hosts).

I might need to raise a totally new bug for that depending on what's been reported already.

The other issue of the multiple default routes prevented successful QA of the patch without manual intervention to delete the extra routes. If we're adverse to multiple fixes in a PR, I can put up a separate fix.

I'll take another look at the other repro scenario you mention.

**Comment by wallyworld** [comment]
Replaced by individual PRs
https://github.com/juju/juju/pull/21548
https://github.com/juju/juju/pull/21549
